### ISSUE TITLE
[Feature Request] Add Stopping Criteria support

### ISSUE THREAD

**Issue by Butanium** [TransformerLensOrg/TransformerLens#595]
Hey,

Huggingface offers a useful [Stopping Criteria class](https://huggingface.co/docs/transformers/internal/generation_utils#transformers.StoppingCriteria) that can be utilized in the `.generate` function to halt generation based on custom conditions.

Do you think it would be possible to integrate this functionality into `HookedTransformer.generate` as well?

**Comment by bryce13950** [comment]
I think this could be a very useful thing to add

**Comment by srishti-git1110** [comment]
Hi @bryce13950 , if this isn't being actively worked on, I'd like to attempt this.

**Comment by RecreationalMath** [comment]
The bridge `generate()` is a custom loop with EOS-only stopping. `hf_generate()` already supports `stop_strings` and `stopping_criteria` natively via `**generation_kwargs`, but it forwards to `self.original_model.generate(...)` and so does not fire TransformerLens hook points per step.

As I understand it, the request is to add the same parameters (`stop_strings` and `stopping_criteria`) to the custom `generate()`, so that the bridge's hook-firing `generate()` can also stop on strings and callables.

Any inputs, @jlarson4?

**Comment by jlarson4** [comment]
@RecreationalMath Yes, that is the same interpretation I'd make of this feature request. Are you interested in taking this on?

**Comment by RecreationalMath** [comment]
Yes, and thanks for the confirmation.
I will send a PR shortly.

**Comment by RecreationalMath** [comment]
Correction to my earlier comment: `hf_generate()` does fire TransformerLens hooks.
